org.lflang/src/org/lflang/generator/ts/TSGenerator.kt

```diff
@@ -245,6 +242,50 @@ class TSGenerator(
         }
     }
 
+
+    /**
+     * If a main or federated reactor has been declared, create a ReactorInstance of it.
+     * This will assign levels to reactions; then, if the program is federated,
+     * an AST transformation is performed to disconnect connections between federates.
+     */
+    private fun createMainReactorInstance() {
+        if (mainDef != null) {
+            if (main == null) {
+                // Recursively build instances. This is done once because
+                // it is the same for all federates.
+                main = ReactorInstance(
```

Comment:
Why is ReactorInstance instantiated here? This could be very expensive.

Reply:
I see that it's needed for `removeRemoteFederateConnectionPorts`. Would it work to limit the depth to 1, like it is done [here](https://github.com/lf-lang/lingua-franca/blob/e3c7fb439ec67b721aba6fdd468d36f58035836b/org.lflang/src/org/lflang/generator/GeneratorBase.java#L1121)?

Reply:
This is just following the implementation in the CGenerator.java here: https://github.com/lf-lang/lingua-franca/blob/master/org.lflang/src/org/lflang/generator/c/CGenerator.java#L2628

Would you have any suggestion on where the ReactorInstance is instantiated instead?

Reply:
The comment above was a reply to the first comment, by the way. 

> I see that it's needed for removeRemoteFederateConnectionPorts. Would it work to limit the depth to 1, like it is done [here](https://github.com/lf-lang/lingua-franca/blob/e3c7fb439ec67b721aba6fdd468d36f58035836b/org.lflang/src/org/lflang/generator/GeneratorBase.java#L1121)?

Sure, I can try that. Do you think we should update the CGenerator's logic as well?

Reply:
I added createMainReactorInstance() to make `findOutputsConnectedToPhysicalActions()` work. So now TypeScript starts using ReactorInstance.

Reply:
Specifically, by phantom connections, I'm referring to the bold lines in the example below that is part of #608:

![](https://user-images.githubusercontent.com/4221770/138773619-24bdc48e-d8df-42a8-b1ed-908695f0964d.png)

Reply:
> Ah, I see. Okay. That makes sense. I got my answer then. There is no need to modify anything then. Thanks for explaining this!

Oh, I see. Thanks for the clarification and review! I'll leave this as is now then. 👍